rook/pool.py

```diff
@@ -0,0 +1,175 @@
+      self.__done = True
+      return
+
+  def is_done(self):
+    """
+    Returns true if this is done running.
+    """
+    return self.__done
+
+class MultiRun:
+  """
+  This creates queues and runner threads to process the functions.
+  """
+  def __init__(self, function_list, number_jobs, ready_to_run=None):
+    """
+    Initializes the class
+    function_list: list of functions and data to run
+    number_jobs: number of functions to run simultaneously.
+    ready_to_run: list of if functions are ready to run
+    """
+    self.__function_list = function_list
+    self.__runners = [None]*number_jobs
+    self.__input_queue = queue.Queue()
+    self.__output_queue = queue.Queue()
+    if ready_to_run is not None:
+      assert len(ready_to_run) == len(self.__function_list)
+      self.__ready_to_run = ready_to_run[:]
+    else:
+      self.__ready_to_run = [True]*len(self.__function_list)
+    self.__not_ready = 0
+
+
+  def run(self):
+    """
+    Starts running all the tests
+    """
+    self.__not_ready = 0
+    for id_num, (function, data) in enumerate(self.__function_list):
+      if self.__ready_to_run[id_num]:
+        self.__input_queue.put((id_num, function, data))
+      else:
+        self.__not_ready += 1
+    for i in range(len(self.__runners)):
+      self.__runners[i] = RunnerThread(self.__input_queue, self.__output_queue)
+    for runner in self.__runners:
+      runner.start()
+
+  def enable_job(self, id_num):
+    """
+    Enables the previously not ready job
+    """
+    assert not self.__ready_to_run[id_num]
+    self.__not_ready -= 1
+    self.__ready_to_run[id_num] = True
+    function, data = self.__function_list[id_num]
+    self.__input_queue.put((id_num, function, data))
+    self.__restart_runners()
+
+  def __restart_runners(self):
+    """
+    Restarts any dead runners
+    """
+    for i in range(len(self.__runners)):
+      if self.__runners[i].is_done() or not self.__runners[i].is_alive():
+        #Restart since it is done.  Otherwise might not be any runners still
+        # running
+        self.__runners[i] = RunnerThread(self.__input_queue, self.__output_queue)
+        self.__runners[i].start()
+
+  def __runner_count(self):
+    """
+    Returns how many runners are not done.
+    """
+    runner_count = 0
+    for i in range(len(self.__runners)):
+      if self.__runners[i].is_alive():
+        runner_count += 1
+    return runner_count
+
+  def process_results(self, process_function=None):
+    """
+    Process results and return the output in an array.
+    If a process_function is passed in, it will be called with
+    process_function(index, input, output) after the output is created.
+    """
+    return_array = [None]*len(self.__function_list)
+    output_count = 0
+    count_down = 10
+    while output_count < len(return_array):
+      while output_count + self.__not_ready < len(return_array):
+        print("meditation numbers oc", output_count, "nr", self.__not_ready,
```

Comment:
Debug or meant to be included?

Reply:
Debug, but needs to be on for the time being because otherwise I may not be able to figure out why the test system crashed randomly.  I'll add a comment about that.

Reply:
This is for debugging, but since the it is meant to help figure out why deadlocks are occurring, it probably needs to be there for the time being.  After we have used the testing system for a while if it doesn't cause problems, we can comment it out.